src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -2166,36 +2222,346 @@ private MemberDeclarationSyntax ParseMemberDeclarationOrStatementCore(SyntaxKind
+                    // an async method.  In that case we reconsider "async" to be a modifier, and treat the identifier
+                    // (with the type parameters) as the type (with type arguments).  Then we go back to looking for
+                    // the member name again.
+                    // For example, if we get
+                    //     async Task<
+                    // then we want async to be a modifier and Task<MISSING> to be a type.
+                    if (ReconsideredTypeAsAsyncModifier(ref modifiers, ref type, ref afterTypeResetPoint, ref explicitInterfaceOpt, ref identifierOrThisOpt, ref typeParameterListOpt))
+                    {
+                        goto parse_member_name;
+                    }
+
+                    Debug.Assert(identifierOrThisOpt != null);
+
+                    // check availability of readonly members feature for indexers, properties and methods
+                    CheckForVersionSpecificModifiers(modifiers, SyntaxKind.ReadOnlyKeyword, MessageID.IDS_FeatureReadOnlyMembers);
+
+                    if (TryParseIndexerOrPropertyDeclaration(attributes, modifiers, type, explicitInterfaceOpt, identifierOrThisOpt, typeParameterListOpt, out result))
+                    {
+                        return result;
+                    }
+
+                    // treat anything else as a method.
+
+                    // PROTOTYPE(SimplePrograms): Allow attributes on local functions
+                    if (!IsScript && !haveAttributes && explicitInterfaceOpt is null &&
+                        tryParseLocalDeclarationStatementFromStartPoint<LocalFunctionStatementSyntax>(ref startPoint, out result))
+                    {
+                        return result;
+                    }
+
+                    return this.ParseMethodDeclaration(attributes, modifiers, type, explicitInterfaceOpt, identifierOrThisOpt, typeParameterListOpt);
+                }
+                finally
+                {
+                    this.Release(ref afterTypeResetPoint);
+                }
+            }
+            finally
+            {
+                _pool.Free(modifiers);
+                _termState = saveTermState;
+                this.Release(ref startPoint);
+            }
+
+            bool tryParseLocalDeclarationStatement<DeclarationSyntax>(out MemberDeclarationSyntax result) where DeclarationSyntax : StatementSyntax
+            {
+                bool wasInAsync = IsInAsync;
+                IsInAsync = true; // We are implicitly in an async context
+                int lastTokenPosition = -1;
+                IsMakingProgress(ref lastTokenPosition);
+
+                var topLevelStatement = ParseLocalDeclarationStatement();
+                IsInAsync = wasInAsync;
+
+                if (topLevelStatement is DeclarationSyntax declaration && IsMakingProgress(ref lastTokenPosition, assertIfFalse: false))
```

Comment:
This feels like a strange way to detect whether the statement that was parsed was meaningful. Could we maybe extract out a helper for ParseLocalDeclarationStatement that returns null when it has nothing to consume from the source text, and the ParseLocalDeclarationStatement method can call it and replace with an "empty" LocalDeclarationStatement to preserve behavior for existing callers?

Reply:
> This feels like a strange way to detect whether the statement that was parsed was meaningful. Could we maybe extract out a helper for ParseLocalDeclarationStatement that returns null when it has nothing to consume from the source text, and the ParseLocalDeclarationStatement method can call it and replace with an "empty" LocalDeclarationStatement to preserve behavior for existing callers?

I think that would still require checking if we are making any progress, but in a different place. I'd rather keep this logic here, to avoid people calling the wrong helper unintentionally. 

---
In reply to: [382867655](https://github.com/dotnet/roslyn/pull/41706#discussion_r382867655) [](ancestors = 382867655)